tortoise/state.go

```diff
@@ -5,78 +5,242 @@ import (
-	// last evicted layer
-	evicted types.LayerID
-
-	// localThreshold is changed when the last layer is updated.
-	// computed as config.LocalThreshold * epochWeights[last].
-	//
-	// used as a part of threshold and for voting when block is undecided and outside of hdist.
-	localThreshold util.Weight
-	// globalThreshold is changed when the last or verified layer is updated.
-	// computed as - sum of all layers between verified + 1 up to last * config.GlobalThreshold + local threshold.
-	globalThreshold util.Weight
-
-	blocks      map[types.LayerID][]blockInfo
-	ballots     map[types.LayerID][]ballotInfo
-	ballotLayer map[types.BallotID]types.LayerID
-	blockLayer  map[types.BlockID]types.LayerID
-
-	// cache ref ballot by epoch
-	refBallotBeacons map[types.EpochID]map[types.BallotID]types.Beacon
-	// cache ballots with bad beacons. this cache is mainly for self-healing where we only have BallotID
-	// in the opinions map to work with.
-	badBeaconBallots map[types.BallotID]struct{}
-
-	// epochWeight average weight per layer of atx's that target keyed epoch
-	epochWeight map[types.EpochID]util.Weight
-	// referenceHeight is a median height from all atxs that target keyed epoch
-	referenceHeight map[types.EpochID]uint64
-
-	// referenceWeight stores atx weight divided by the total number of eligibilities.
-	// it is computed together with refBallot weight. it is not equal to refBallot
-	// only if refBallot has more than 1 eligibility proof.
-	referenceWeight map[types.BallotID]util.Weight
-
-	hareOutput map[types.LayerID]types.BlockID
-	validity   votes
-}
-
-type blockInfo struct {
-	id     types.BlockID
-	height uint64
-	weight util.Weight
-}
-
-type ballotInfo struct {
-	id     types.BallotID
-	height uint64
-	weight util.Weight
+type (
+	weight = util.Weight
+
+	verifyingInfo struct {
+		good, abstained weight
+		referenceHeight uint64
+	}
+
+	layerInfo struct {
+		lid            types.LayerID
+		empty          weight
+		hareTerminated bool
+		blocks         []*blockInfo
+		verifying      verifyingInfo
+	}
+
+	blockInfo struct {
+		id       types.BlockID
+		layer    types.LayerID
+		height   uint64
+		hare     sign
+		validity sign
+		margin   weight
+	}
+
+	state struct {
+		// last received layer
+		// TODO should be last layer according to the clock
+		last types.LayerID
+		// last verified layer
+		verified types.LayerID
+		// historicallyVerified matters only for local opinion for verifying tortoise
+		// during rerun. for live tortoise it is identical to the verified layer.
+		historicallyVerified types.LayerID
+		// last processed layer
+		minprocessed, processed types.LayerID
+		// last evicted layer
+		evicted types.LayerID
+
+		localThreshold  weight
+		globalThreshold weight
+
+		// epochWeight average weight per layer of atx's that target keyed epoch
+		epochWeight map[types.EpochID]weight
+		// referenceHeight is a median height from all atxs that target keyed epoch
+		referenceHeight map[types.EpochID]uint64
```

Comment:
given that the goal is trying to be memory efficient, maybe it's worth consolidating maps with the same key type.
i.e. combine epochWeight/referenceHeight to be a map[types.EpochID]epochInfo that has two uint64

same for maps with LayerID/BallotID

Reply:
i tried that and i have to refactor even more tests, i would rather do it some time in the future. having those things as they are doesn't really affect memory much